File: src/main/groovy/jokatu/components/controllers/game/MessageHistoryController.java
package jokatu.components.controllers.game;

import jokatu.components.stores.MessageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.Message;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;
import java.util.List;

import static org.springframework.web.bind.annotation.RequestMethod.GET;

@RestController
public class MessageHistoryController {

	private final MessageRepository messageRepository;

	@Autowired
	public MessageHistoryController(MessageRepository messageRepository) {
		this.messageRepository = messageRepository;
	}

	@RequestMapping(value = "/messages", method = GET)
	List<Message> getMessages(@RequestParam("destination") String destination, Principal principal) {
		if (destination.startsWith("/user")) {
			return messageRepository.getMessageHistoryForUser(principal.getName(), destination.substring(5));
		} else {
			return messageRepository.getMessageHistory(destination);
		}
	}

	@RequestMapping(value = "/last_message", method = GET)
	Message getLastMessage(@RequestParam("destination") String destination, Principal principal) {
		if (destination.startsWith("/user")) {
			return messageRepository.getLastMessageForUser(principal.getName(), destination.substring(5));
		} else {
			return messageRepository.getLastMessage(destination);
		}
	}
}

File: src/main/kotlin/jokatu/components/controllers/game/MessageHistoryController.kt
package jokatu.components.controllers.game

import jokatu.components.stomp.UnknownDestination
import jokatu.components.stores.MessageRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.messaging.Message
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod.GET
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.security.Principal

/**
 * A REST controller for getting the history and last message sent to a STOMP channel.
 */
@RestController
class MessageHistoryController @Autowired constructor(private val messageRepository: MessageRepository) {

	@RequestMapping(value = ["/messages"], method = [GET])
	internal fun getMessages(@RequestParam("destination") destination: String, principal: Principal): List<Message<*>> {
		return if (destination.startsWith("/user")) {
			messageRepository.getMessageHistoryForUser(principal.name, UnknownDestination(destination.substring(5)))
		} else {
			messageRepository.getMessageHistory(UnknownDestination(destination))
		}
	}

	@RequestMapping(value = ["/last_message"], method = [GET])
	internal fun getLastMessage(@RequestParam("destination") destination: String, principal: Principal): Message<*>? {
		return if (destination.startsWith("/user")) {
			messageRepository.getLastMessageForUser(principal.name, UnknownDestination(destination.substring(5)))
		} else {
			messageRepository.getLastMessage(UnknownDestination(destination))
		}
	}
}
